src/Compiler/AbstractIL/ilread.fs

```diff
@@ -2107,7 +2107,37 @@ and typeDefReader ctxtH : ILTypeDefStored =
         let struct (endFieldsIdx, endMethodsIdx) = seekReadTypeDefRowExtents ctxt info idx
         let typars = seekReadGenericParams ctxt 0 (tomd_TypeDef, idx)
         let numTypars = typars.Length
-        let super = seekReadOptionalTypeDefOrRef ctxt numTypars AsObject extendsIdx
+
+        let kind =
+            let extendsTag = extendsIdx.tag
+            let extendsIdx = extendsIdx.index
+
+            if flags &&& 0x00000020 <> 0x0 then
+                ILTypeDefAdditionalFlags.Interface
+            else if extendsIdx = 0 && extendsTag = tdor_TypeDef then
+                ILTypeDefAdditionalFlags.Class
+            else
+                let extendsName =
+                    if extendsTag = tdor_TypeDef then
+                        let mutable addr = ctxt.rowAddr TableNames.TypeDef extendsIdx
+                        let _ = seekReadInt32Adv mdv &addr
+                        let nameIdx = seekReadStringIdx ctxt mdv &addr
+                        let namespaceIdx = seekReadStringIdx ctxt mdv &addr
+                        readBlobHeapAsTypeName ctxt (nameIdx, namespaceIdx)
+                    elif extendsTag = tdor_TypeRef then
+                        let _, nameIdx, namespaceIdx = seekReadTypeRefRow ctxt mdv extendsIdx
+                        readBlobHeapAsTypeName ctxt (nameIdx, namespaceIdx)
+                    else
+                        ""
+
+                match extendsName with
```

Comment:
Is this only approximate?
If yes, what cases are missing?

Reply:
The main idea here is to seek read from super type only name and check it as in `typeKindOfFlags` https://github.com/dotnet/fsharp/blob/f2fe83295721c91d1a1f2372d84bc737f7660f77/src/Compiler/AbstractIL/il.fs#L2580-L2592

Super type can be `TypeDef`/`TypeRef` or `TypeSpec`. 
`TypeSpec` is not our case because (please correct if I'm wrong) `System.Enum`/`System.Delegate`/`System.MulticastDelegate/System.ValueType` is not an instantiation of a generic type and cannot be the `TypeSpec`, so we could just check that super type is `TypeDef`/`TypeRef` and seek read its name in a corresponding way.

Reply:
I trust the change in general, but we did have regressions (e.g. this year for .NET Interactive) caused by slight changes in ilreading -> since its all plain string literals, it really is easy to make a mistake.

e.g. for various delegates created from C# compilations, it would be good to have small tests guarding that the behavior is the same.

Reply:
https://github.com/dotnet/fsharp/pull/18005/commits/1a886f5dfc8c3e29c90f96ce06b6e8c735746b16